R/validate_links.R

```diff
@@ -0,0 +1,185 @@
+#'
+#' These links will have no protocol, but should resolve to the HTML version of
+#' a page and have the correct capitalisation.
+#'
+#' ## Anchors (aka fragments)
+#'
+#' Anchors are located at the end of URLs that start with a `#` sign. These are
+#' used to indicate a section of the documenation.
+#'
+#' ## Alt-text (for images)
+#'
+#' All images must have associated alt-text. In pandoc, this is acheived by
+#' writing the `alt` attribute in curly braces after the image: 
+#' `![image caption](link){alt='alt text'}`
+#'
+#' ## Descriptive text
+#'
+#' All links must have descriptive text associated with them, which is 
+#' beneficial for screen readers scanning the links on a page to not have a
+#' list full of "link", "link", "link".
+#'
+#' @param yrn a [tinkr::yarn] or [Episode] object.
+#' @param verbose if TRUE (default), messages will be printed as the validator works.
+#' @return a data frame containing the links, locations, and indicators if they
+#' passed tests
+validate_links <- function(yrn, verbose = TRUE) {
+  has_cli <- is.null(getOption("pegboard.no-cli")) &&
+    requireNamespace("cli", quietly = TRUE)
+
+  link_table <- make_link_table(yrn)
+  path <- basename(yrn$path)
+  VLD <- c(
+    enforce_https = TRUE,
+    internal_okay = TRUE,
+    all_reachable = TRUE,
+    img_alt_text  = TRUE,
+    descriptive   = TRUE,
+    NULL
+  )
+  VLD["descriptive"]   <- validate_descriptive(link_table, path, verbose, has_cli)
+  VLD["enforce_https"] <- validate_https(link_table, path, verbose, has_cli)
+  VLD["img_alt_text"]  <- validate_alt_text(link_table, path, verbose, has_cli)
+  VLD["internal_okay"] <- validate_internal_okay(yrn, link_table, verbose, has_cli)
+  VLD
+}
+
+validate_descriptive <- function(lt, path, verbose = TRUE, cli = TRUE) {
+  res <- !any(bad <- tolower(lt$text) %in% c("link", "this link", "a link"))
+  if (verbose && !res) {
+    just_link    <- lt[bad, , drop = FALSE]
+    link_sources <- glue::glue("{path}:{just_link$sourcepos}")
+    issue_warning("Link text should be more descriptive than {sQuote('link')}:
+      {glue::glue_collapse(lnks, sep = '\n')}",
+      cli,
+      lnks = glue::glue("{sQuote(just_link$text)} ({link_sources})")
+    )
+  }
+  res
+}
+
+validate_internal_okay <- function(yrn, lt = NULL, verbose = TRUE, cli = TRUE) {
+  if (is.null(lt)) {
+    lt <- make_link_table(yrn)
+  }
+  headings   <- xml2::xml_text(yrn$headings)
+  path       <- basename(yrn$path)
+  internal   <- lt$server == "" & lt$scheme == "" & is.na(lt$port) & lt$user == ""
+  cross_page <- internal & lt$path != ""
+  in_page    <- internal & lt$path == "" & lt$fragment != ""
+  res <- TRUE
+  if (any(in_page)) {
+    our_headings <- clean_headings(headings)
+    no_match <- !lt$fragment %in% our_headings
+    res <- res & !any(in_page & no_match)
+  } else {
+    no_match <- FALSE
+  }
+  if (any(cross_page)) {
+    neighbors <- get_lesson_files(yrn$lesson)
+    no_file <- !lt$path %in% glue::glue("{neighbors}")
```

Comment:
The comparison of the paths here means that image path written as `./img/file.png` for instance, are marked as missing.

Reply:
I've addressed this in 08d7746